cpp/src/io/fst/agent_dfa.cuh

```diff
@@ -538,6 +756,37 @@ __launch_bounds__(int32_t(AgentDFAPolicy::BLOCK_THREADS)) CUDF_KERNEL
   // The state transition vector passed on to the second stage of the algorithm
   StateVectorT out_state_vector;
 
+  using OutSymbolT = typename DfaT::OutSymbolT;
+  // static constexpr int32_t MIN_TRANSLATED_OUT = DfaT::MIN_TRANSLATED_OUT;
+  static constexpr int32_t num_max_translated_out = DfaT::MAX_TRANSLATED_OUT;
+  static constexpr bool discard_out_index =
+    ::cuda::std::is_same<TransducedIndexOutItT, thrust::discard_iterator<>>::value;
+  static constexpr bool discard_out_it =
+    ::cuda::std::is_same<TransducedOutItT, thrust::discard_iterator<>>::value;
+  using NonWriteCoalescingT =
+    DFASWriteCallbackWrapper<num_max_translated_out,
+                             decltype(dfa.InitTranslationTable(transducer_table_storage)),
+                             TransducedOutItT,
+                             TransducedIndexOutItT>;
+
+  using WriteCoalescingT =
+    WriteCoalescingCallbackWrapper<discard_out_index,
+                                   discard_out_it,
+                                   num_max_translated_out * SYMBOLS_PER_BLOCK,
+                                   OutSymbolT,
+                                   decltype(dfa.InitTranslationTable(transducer_table_storage)),
+                                   TransducedOutItT,
+                                   TransducedIndexOutItT>;
+
+  static constexpr bool is_translation_pass = (!IS_TRANS_VECTOR_PASS) || IS_SINGLE_PASS;
+
+  // Use write-coalescing only if the worst-case output size per tile fits into shared memory
+  static constexpr bool use_shmem_cache =
+    is_translation_pass and (sizeof(typename WriteCoalescingT::TempStorage) <= 24 * 1024);
```

Comment:
Question: Since 24KB per block is a heuristic, can we add it to the tuning policy in `dispatch_dfa`?

Reply:
That's a great idea. I added it as a parameter on the tuning policy and also added another check that makes sure that the total static shared memory allocation doesn't exceed the per-CTA static smem allocation limit of 48 KiB.